File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/FleetApiClient.swift
import Foundation

/// Minimal REST client aligned with `docs/openapi/fleet-api.yaml`.
public final class FleetApiClient {
    private let options: FleetSdkOptions
    private let session: URLSession

    public init(options: FleetSdkOptions, session: URLSession = .shared) {
        self.options = options
        self.session = session
    }

    public static func `default`(resolvedOptions: FleetSdkOptions?) throws -> FleetApiClient {
        guard let o = resolvedOptions else {
            throw FleetSdkError(code: .notInitialized, message: "FleetSdk.initialize required.")
        }
        return FleetApiClient(options: o)
    }

    public func listDrivers(completion: @escaping (Result<[Driver], Error>) -> Void) {
        if options.useMock {
            completion(.success([
                Driver(
                    id: "DRV001",
                    name: "Demo Driver",
                    vrn: "MH 02 AB 1234",
                    status: "Active",
                    cardBalancePaise: 1_250_000,
                ),
            ]))
            return
        }

        var base = options.apiBaseUrl
        while base.hasSuffix("/") {
            base.removeLast()
        }
        guard let url = URL(string: base + "/fleet/drivers") else {
            completion(.failure(FleetSdkError(code: .invalidInput, message: "Bad apiBaseUrl")))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("mgl-fleet-ios-sdk/1.0", forHTTPHeaderField: "User-Agent")
        if let token = options.authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        session.dataTask(with: request) { data, response, error in
            if let error {
                completion(.failure(error))
                return
            }
            guard let http = response as? HTTPURLResponse else {
                completion(.failure(FleetSdkError(code: .networkError, message: "No HTTP response")))
                return
            }
            let bodyText = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
            guard (200 ..< 300).contains(http.statusCode) else {
                let flat = bodyText.replacingOccurrences(of: "\n", with: " ").trimmingCharacters(in: .whitespaces)
                let snippet = String(flat.prefix(200))
                let htmlHint =
                    flat.trimmingCharacters(in: .whitespaces).hasPrefix("<")
                    ? " [Non-JSON HTML — check apiBaseUrl, auth/WAF, or use useMock:true.]"
                    : ""
                completion(.failure(FleetSdkError(
                    code: .networkError,
                    message: "HTTP \(http.statusCode) \(url.absoluteString): \(snippet)\(htmlHint)",
                )))
                return
            }
            guard let data else {
                completion(.failure(FleetSdkError(code: .networkError, message: "Empty response")))
                return
            }
            do {
                let decoded = try JSONDecoder().decode(DriversResponse.self, from: data)
                completion(.success(decoded.drivers))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    private struct DriversResponse: Codable {
        let drivers: [Driver]
    }
}

File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/api/FleetApiClient.kt
package com.mgl.fleet.sdk.api

import com.mgl.fleet.sdk.FleetSdkHolder
import com.mgl.fleet.sdk.FleetSdkOptions
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Minimal REST client aligned with [fleet-api.yaml](/docs/openapi/fleet-api.yaml).
 */
class FleetApiClient(
    private val options: FleetSdkOptions,
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build(),
) {

    fun listDrivers(): Result<List<Driver>> {
        if (options.useMock) {
            return Result.success(
                listOf(
                    Driver(
                        id = "DRV001",
                        name = "Demo Driver",
                        vrn = "MH 02 AB 1234",
                        status = "Active",
                        cardBalancePaise = 1_250_000L,
                    ),
                ),
            )
        }
        return getJson("/fleet/drivers") { json ->
            val arr = json.getJSONArray("drivers")
            buildList {
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    add(parseDriver(o))
                }
            }
        }
    }

    private fun parseDriver(o: JSONObject): Driver =
        Driver(
            id = o.getString("id"),
            name = o.getString("name"),
            vrn = o.getString("vrn"),
            status = o.getString("status"),
            cardBalancePaise = o.getLong("cardBalancePaise"),
        )

    private fun <T> getJson(path: String, parse: (JSONObject) -> T): Result<T> {
        val base = options.apiBaseUrl.trimEnd('/')
        val reqBuilder =
            Request.Builder()
                .url("$base$path")
                .get()
                .header("Accept", "application/json")
                .header("User-Agent", "mgl-fleet-android-sdk/1.0")

        options.authToken?.let { reqBuilder.header("Authorization", "Bearer $it") }
        val req = reqBuilder.build()
        return try {
            client.newCall(req).execute().use { resp ->
                val body = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) {
                    val flat = body.replace("\n", " ").trim()
                    val snippet = flat.take(200).let { if (flat.length > 200) "$it…" else it }
                    val htmlHint =
                        if (flat.startsWith("<"))
                            " [Non-JSON HTML — wrong apiBaseUrl, missing auth/WAF blocking, or no GET /fleet/drivers. Dev: set useMock: true.]"
                        else
                            ""
                    return Result.failure(
                        IOException(
                            "HTTP ${resp.code} ${resp.request.url}: ${snippet.ifEmpty { "(empty)" }}$htmlHint",
                        ),
                    )
                }
                Result.success(parse(JSONObject(body)))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    companion object {
        /** Resolved client using globally initialized options (throws if unset). */
        fun default(): FleetApiClient = FleetApiClient(FleetSdkHolder.requireOptions())
    }
}
